### PR TITLE
[BEAM-5265] Use currentProcessingTime() for onTime with processing time domain

### PR DESCRIPTION
**Pull request by JozoVilcek**
Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by JozoVilcek** [comment]
Simple patch to see if it breaks any existing tests

**Comment by swegner** [comment]
Thanks for the contribution @JozoVilcek. Is this ready for review? If so, please assign it to a reviewer by leaving a comment with:

```
R: @reviewerName
```

If you need help finding a reviewer, you can check in the git commit history or component [OWNERS file](https://github.com/apache/beam/blob/master/runners/core-java/OWNERS). See: https://s.apache.org/beam-owners

**Comment by JozoVilcek** [comment]
R: @kennknowles

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java [hunk 1]
```diff
@@ -256,6 +256,8 @@ public void onTimer(
         break;
 
       case PROCESSING_TIME:
+        effectiveTimestamp = stepContext.timerInternals().currentProcessingTime();
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java]: Ah, there is a general problem here. The `effectiveTimestamp` is an event time timestamp for the elements that are output from the `@OnTimer` method. The current processing time is in a different time domain that may be different in arbitrary ways.

**Comment by kennknowles** [comment]
It is a good point that we should have a test that makes this fail - would you find it interesting to try to write this test? You should be able to do it using `TestStream`.

**Comment by JozoVilcek** [comment]
Sure, will do. As stated in JIRA, I was first trying to find out if fix around here does makes sense.

**Comment by JozoVilcek** [comment]
@kennknowles I have updated tests, hopefully on the right path. I am sorry for the delay.

**Comment by JozoVilcek** [comment]
I am not sure if I am up to the fun :) Can not tell because I am not clear on what should be the test like. Which records can be dropped because of this change? My understanding of internals is rather shallow :/

**Comment by kennknowles** [comment]
The processing time can be arbitrarily far ahead of the watermark, or behind it. They are independent, even though in a lot of streaming applications they will be near each other.

So if the watermark is behind processing time, you'll get timestamps that are outside the window and that is forbidden. If the watermark is ahead of processing time (more rare and weird probably, since you would be processing future events) then you get dropping.

**Comment by JozoVilcek** [comment]
Yes, makes sense.
So in code it essentially means do a similar test case with IntervalWindow not just global and play with processing time being before / after watermark and inside / outside window boundary + outside of allowed lateness. Correct?

**Comment by JozoVilcek** [comment]
Maybe before / after watermark does not matter that much

**Comment by kennknowles** [comment]
Yes, that's exactly what I was thinking of

**Comment by JozoVilcek** [comment]
I am not sure if result of tests are in line with your expectations. It seems that processing time being out of window of fine as long as watermark is within the window and it's lateness.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java [hunk 2]
```diff
@@ -198,44 +218,185 @@ public void testFinishBundleExceptionsWrappedAsUserCodeException() {
...
+                window.start().plus(2 * DoFnWithTimers.TIMER_OFFSET.getMillis()).plus(2))
+
+            // watermark in window, processing time ahead of window.end() but within lateness
+            .advanceProcessingTime(windowDuration)
+            .addElements(event)
+            .advanceProcessingTime(DoFnWithTimers.TIMER_OFFSET.plus(1))
+            .advanceWatermarkTo(
+                window.start().plus(3 * DoFnWithTimers.TIMER_OFFSET.getMillis()).plus(3))
+
+            // watermark in window, processing time  is out of window's allowed lateness
```
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: If this doesn't cause a crash, then we have a bug!

Comment (JozoVilcek) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: So, what now. I have no idea where to look (did not try yet to be frank) :) Which part decides on valid / dropped elements?

Comment (JozoVilcek) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Hey @kennknowles , any comments? Should I add dropping at SimpleDoFnRunner?

Comment (kennknowles) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: I think a more isolated test case might be easier. You want processing time to be way beyond the allowed lateness. Then when you emit elements I think `outputWithTimestamp` should actually crash saying that the timestamp was out of bounds.

Comment (JozoVilcek) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: What kind of isolation do you have in mind? I was under the impression that if one wants to processElements, set timer and get timer callback called, test pipeline is needed. Is there example I can follow?

I am still unclear on what needs to be dropped when by who. Dynamics of DoFnRunner is quite complex. Test case does not use `outputWithTimestamp` but only `output`. Both just constructs  `WindowedValue` and fire them away.

Also, there are no limitations on for when timer can be set, but there are some event time.
https://github.com/apache/beam/blob/master/runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java#L995

Comment (JozoVilcek) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Without further input, I can not finish this. Sorry

**Pull request closed without merging**